File: python/PStest/2020_k/4.py
import time

prev_time = time.time_ns()

class Node:
    def __init__(self, key):
        self.key = key
        self.cnt = 0
        self.child = {}
        self.isTerminated = False

class Trie:
    def __init__(self):
        self.head = Node(None)

    def insert(self, s):
        curr_node = self.head

        for c in s:
            if c not in curr_node.child:
                curr_node.child[c] = Node(c)
            curr_node.cnt += 1
            curr_node = curr_node.child[c]
        curr_node.isTerminated = True

    def count(self, query):
        cnt = 0
        cur_node = self.head
        for i in range(len(query)):
            if query[i] == '?':
                cnt = cur_node.cnt
                return cnt
            if query[i] in cur_node.child:
                cur_node = cur_node.child[query[i]]
            else:
                return 0
        return -1

def solution(words, queries):
    answer = []

    t = [Trie() for _ in range(10001)]
    rt = [Trie() for _ in range(10001)]
    for w in words:
        t[len(w)].insert(w)
        rt[len(w)].insert(w[::-1])
    for query in queries:
        cnt = 0
        if query[0] != '?':
            q = query
            cnt = t[len(q)].count(q)
        else:
            q = query[::-1]
            cnt = rt[len(q)].count(q)
        answer.append(cnt)

    return answer


words = ["frodo", "front", "frost", "frozen", "frame", "kakao", "a"]
# queries = ["fro??", "????o", "fr???", "fro???", "pro?"]
queries = ["frodo?"]

print(solution(words,queries))

print('time', time.time_ns()-prev_time)

# test_str = 'frodo'
# print(test_str,t.find(test_str))
# test_str = 'f'
# print(test_str,t.find(test_str))
# test_str = 'front'
# print(test_str,t.find(test_str))
# test_str = '?ront'
# print(test_str,t.find(test_str))
# test_str = 'frod?'
# print(test_str,t.find(test_str))
# test_str = 'froe?'
# print(test_str,t.find(test_str))

File: programmers/kakao 2020/4.cpp
//
// Created by Junho Kim on 2021/09/09.
//

#include <iostream>




#include <string>
#include <vector>

using namespace std;

struct Trie;
Trie* alloc();

struct Trie {
    Trie *child[26];
    int cnt;

    void add(char *str) {
        if (*str == 0) {
            cnt++;
            return;
        }
        int idx = *str - 'a';
        if (idx >= 26 || idx < 0) {
            idx = 26;
        }
        if (child[idx] == nullptr) {
            child[idx] = alloc();
        }
        child[idx]->add(str + 1);
    }

    int count() {
        for (int i = 0; i < 26; ++i) {
            if (child[i]) {
                cnt += child[i]->count();;
            }
        }
        return cnt;
    }

    int find(char *str) {
        if (*str == '?') {
            return cnt;
        }
        int idx = *str - 'a';
        if (!child[idx]) {
            return 0;
        }
        return child[idx]->find(str + 1);
    }

    void clear() {
        for (int i = 0; i < 27; ++i) {
            child[i] = nullptr;
        }
    }
};

Trie *alloc() {
    Trie *trie = new Trie();
    trie->clear();
    trie->cnt = 0;
    return trie;
}

Trie *root[10001];
Trie *rroot[10001];

void reverse(char str[], string s) {
    for (int i = 0; i < s.size(); ++i) {
        str[i] = s[s.size() - 1 - i];
    }
    str[s.size()] = '\0';
}

vector<int> solution(vector<string> words, vector<string> queries) {
    vector<int> answer;
    for (int i = 0; i <= 10000; ++i) {
        root[i] = alloc();
        rroot[i] = alloc();
    }

    char rstr[10001];
    for (int i = 0; i < words.size(); ++i) {
        root[words[i].size()]->add((char*) (words[i].c_str()));
        reverse(rstr, words[i]);
        rroot[words[i].size()]->add(rstr);
    }

    for (int i = 0; i <= 10000; ++i) {
        root[i]->count();
        rroot[i]->count();
    }

    for (int i = 0; i < queries.size(); ++i) {
        int cnt = 0;
        if (queries[i][0] == '?') {
            reverse(rstr, queries[i]);
            cnt = rroot[queries[i].size()]->find(rstr);
        } else {
            cnt = root[queries[i].size()]->find((char*) queries[i].c_str());
        }
        answer.push_back(cnt);
    }

    return answer;
}




using namespace std;

int main() {
    vector<string> words = {"frodo", "front", "frost", "frozen", "frame", "kakao"};
    vector<string> queries = {"fro??", "????o", "fr???", "fro???", "pro?"};

    solution(words, queries);



    return 0;
}








